packages/vector_graphics/lib/vector_graphics.dart

```diff
@@ -1,7 +1,190 @@
-library vector_graphics;
+import 'dart:typed_data';
+import 'dart:ui' as ui;
+import 'dart:io';
 
-/// A Calculator.
-class Calculator {
-  /// Returns [value] plus 1.
-  int addOne(int value) => value + 1;
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+import 'package:vector_graphics_codec/vector_graphics_codec.dart';
+
+import 'src/listener.dart';
+
+const VectorGraphicsCodec _codec = VectorGraphicsCodec();
+
+/// Decode a vector graphics binary asset into a [ui.Picture].
+///
+/// Throws a [StateError] if the data is invalid.
+ui.Picture decodeVectorGraphics(ByteData data) {
+  final FlutterVectorGraphicsListener listener =
+      FlutterVectorGraphicsListener();
+  _codec.decode(data, listener);
+  return listener.toPicture();
+}
+
+/// A widget that displays a vector_graphics formatted asset.
+class VectorGraphics extends StatefulWidget {
+  const VectorGraphics({Key? key, required this.controller}) : super(key: key);
+
+  final VectorGraphicsController controller;
+
+  @override
+  State<VectorGraphics> createState() => _VectorGraphicsWidgetState();
+}
+
+class _VectorGraphicsWidgetState extends State<VectorGraphics> {
+  ui.Picture? _picture;
+
+  @override
+  void initState() {
+    _loadAssetBytes();
+    super.initState();
+  }
+
+  @override
+  void didUpdateWidget(covariant VectorGraphics oldWidget) {
+    if (!oldWidget.controller.equivalent(widget.controller)) {
+      _loadAssetBytes();
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  void dispose() {
+    _picture?.dispose();
+    _picture = null;
+    super.dispose();
+  }
+
+  void _loadAssetBytes() {
+    widget.controller.loadBytes().then((ByteData data) {
+      final ui.Picture picture = decodeVectorGraphics(data);
+      setState(() {
+        _picture?.dispose();
+        _picture = picture;
+      });
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ui.Picture? picture = _picture;
+    if (picture == null) {
+      return const SizedBox();
+    }
+    return _RawVectorGraphicsWidget(picture: picture);
+  }
+}
+
+abstract class VectorGraphicsController {
+  const VectorGraphicsController();
+
+  Future<ByteData> loadBytes();
+
+  bool equivalent(VectorGraphicsController other) => false;
```

Comment:
What's the advantage of this vs. marking this class `@immutable` and expecting subclasses to implement == and hashCode?

Reply:
I don't actually want full equality though. Consider the NetworkImageLoader that accepts an HttpClient. if the user accedentally creates a new http client we don't want to reload the bytes...

But of course, this is all sort of defensive. I could entirely remove the controller code and  say "If you want this to work correctly YOU MUST NOt allocated a new controller in each build method"

Reply:
I mean "entirely remove the equality code"

Reply:
Maybe it'd be good to just have an immutable key that controllers could hold for comparison purposes?